core/src/main/java/org/axonframework/eventhandling/TrackingEventProcessor.java

```diff
@@ -327,6 +327,37 @@ public void pause() {
         this.state.updateAndGet(s -> s.isRunning() ? State.PAUSED : s);
     }
 
+    /**
+     * Indicates whether this tracking processor can construct a new {@code Thread} to process
+     * another segment of events.
+     *
+     * @return {@code true} if this tracking processor can construct a new {@code Thread}, {@code false} otherwise
+     */
+    public boolean hasAvailableThreads() {
+        return maxThreadCount > activeProcessorThreads();
+    }
+
+    /**
+     * Blacklist a segment identifier only if it is currently claimed by this instance.
+     * This causes the segment to be released and to no longer to be claimed until it's removed from the black list.
+     *
+     * @param segmentId the id of the segment to be blacklisted
+     */
+    public void blacklistSegment(int segmentId) {
+        if (activeSegments.containsKey(segmentId)) {
+            segmentReleaseDeadlines.put(segmentId, System.currentTimeMillis() + DEFAULT_BACKOFF_TIME_MILLIS * 2);
+        }
+    }
+
+    private boolean canClaimSegment(int segmentId) {
+        if (segmentReleaseDeadlines.containsKey(segmentId) &&
+                segmentReleaseDeadlines.get(segmentId) > System.currentTimeMillis()){
+            return false;
+        }
+        segmentReleaseDeadlines.remove(segmentId);
```

Comment:
There is a small race condition here. If `containsKey()` returns `false` and then an entry is added, this will remove it immediately.
I think it's better to split the `if` into two separate statements and use `remove(Object, Object)` to make sure it's only removed if it's the exact entry we're expecting to remove.

Reply:
Right, I've fixed it as you suggested.
Alternatively we could also avoid removing the entry, always keeping in the map the last deadline for any segment. In this case the method could simply become:

`return !segmentReleaseDeadlines.containsKey(segmentId) || segmentReleaseDeadlines.get(segmentId) < System.currentTimeMillis();`